ADR 009: Modular AnteHandler (#4942)
cosmos/cosmos-sdk: docs/architecture/adr-009-modular-antehandler.md

# ADR 009: Modular AnteHandler

## Changelog

- 2019 Aug 31: Initial draft

## Context

The current AnteHandler design allows users to either use the default AnteHandler provided in `x/auth` or to build their own AnteHandler from scratch. Ideally AnteHandler functionality is split into multiple, modular functions that can be chained together along with custom ante-functions so that users do not have to rewrite common antehandler logic when they want to implement custom behavior.

## Proposals

### Per-Module AnteHandler

One approach is to use the [ModuleManager](https://godoc.org/github.com/cosmos/cosmos-sdk/types/module) and have each module implement its own antehandler if it requires custom antehandler logic. The ModuleManager can then be passed in an AnteHandler order in the same way it has an order for BeginBlockers and EndBlockers. The ModuleManager returns a single AnteHandler function that will take in a tx and run each module's `AnteHandle` in the specified order. The module manager's AnteHandler is set as the baseapp's AnteHandler.

Pros:
1. Simple to implement
2. Utilizes the existing ModuleManager architecture

Cons:

> 1. Improves granularity but still cannot get more granular than a per-module basis. e.g. If auth's `AnteHandle` function is in charge of validating memo and signatures, users cannot swap the signature-checking functionality while keeping the rest of auth's `AnteHandle` functionality.

Author: This would only be possible if the modules implemented their antehandlers with exported micro-functions.

Is this worth noting in the ADR itself?

2. Module AnteHandler are run one after the other. There is no way for one AnteHandler to wrap or "decorate" another.

### Weave Decorators

The Weave approach acheives modularity through the use of Decorators. Weave introduces a new interface:

```golang
// Decorator wraps a Handler to provide common functionality
// like authentication, or fee-handling, to many Handlers
type Decorator interface {
	Check(ctx Context, store KVStore, tx Tx, next Checker) (*CheckResult, error)
	Deliver(ctx Context, store KVStore, tx Tx, next Deliverer) (*DeliverResult, error)
}
```

Each decorator works like a modularized SDK antehandler function, but it can take in a `next` argument that may be another decorator or a Handler (which does not take in a next argument). These decorators can be chained together, one decorator being passed in as the `next` argument of the previous decorator in the chain. The chain ends in a Router which can take a tx and route to the appropriate msg handler.

> A key benefit of this approach is that one Decorator can wrap its internal logic around the next Checker/Deliverer. A weave Decorator may do the following:

Author: One example of where the decorator approach would be useful for us is to have a Decorator that wraps all other AnteHandlers with a defer clause that will catch an `OutOfGas` panic and return the appropriate sdk Result and abort.

In the per-module implementation, it was clear that having a defer clause in every antehandler made no sense. The best way I found to handle this was to set the `ctx.GasMeter` in the module manager's antehandler and then place a defer clause there. However, doing this required changing the `sdk.Tx` interface to include a `Gas` method. 
This can be seen [here](https://github.com/cosmos/cosmos-sdk/blob/aditya/modular-ante/types/module/module.go#L299).

Having a decorator that wraps all subsequent antehandler functions to set the `GasMeter` and implements the defer clause correctly before calling `next` may help us avoid changing the `sdk.Tx` interface. However, any chain that uses something other than `auth.StdTx` will have to implement their own decorator.

```golang
// Example Decorator's Deliver function
func (example Decorator) Deliver(ctx Context, store KVStore, tx Tx, next Deliverer) {
    // Do some pre-processing logic

    res, err := next.Deliver(ctx, store, tx)

    // Do some post-processing logic given the result and error
}
```

Pros:
1. Weave Decorators can wrap over the next decorator/handler in the chain. The ability to both pre-process and post-process may be useful in certain settings.
2. Provides a nested modular structure that isn't possible in the solution above, while also allowing for a linear one-after-the-other structure like the solution above.

Cons:

> 1. It is hard to understand at first glance the state updates that would occur after a Decorator runs given the `ctx`, `store`, and `tx`. A Decorator can have an arbitrary number of nested Decorators being called within its function body, each possibly doing some pre- and post-processing before calling the next decorator on the chain. Thus to understand what a Decorator is doing, one must also understand what every other decorator further along the chain is also doing. This can get quite complicated to understand. A linear, one-after-the-other approach while less powerful, may be much easier to reason about.

Reviewer: I think it is not much harder to reason about.

Decorators are short and should be well-documented.
The only difference with the chained approach is they allow for optional cleanup.
In many cases, they will be `do_check(); next()`, as in chain. But you can add `defer()` clauses to handle panics, or various other cleanups - eg. we can wrap the store to record all keys written, and then auto-add tags *to the result*. 

But I guess "hard to reason about" is quite subjective here. I would recommend you refer to existing weave decorators you find confusing. There is only one I can think of, required by complex business logic from iov.

Reviewer 2: This isn't anything specific to weave. Long chains of middleware/decorator calls do not lend themselves to be easily "readable" -- this is what I think @AdityaSripal meant here and I don't think it's too subjective. We already use this pattern in certain places in the SDK and you have to have some context to understand the execution path.

### Chained Micro-Functions

> The benefit of Weave's approach is that the Decorators can have very narrowly defined functionality, which when chained together allows for maximum customizability. However, the nested structure can get quite complex and thus hard to reason about.

Reviewer 2: ```suggestion
The benefit of Weave's approach is that the Decorators can be very concise, which when chained together allows for maximum customizability. However, the nested structure can get quite complex and thus hard to reason about.
```

> My preferred approach is to split the AnteHandler functionality into tightly scoped "micro-functions", while preserving the one-after-the-other ordering that would come from the per-module approach.

Reviewer 2: ```suggestion
Our recommended approach is to split the AnteHandler functionality into tightly scoped "micro-functions", while preserving the one-after-the-other ordering that would come from the ModuleManager approach.
```

> We can then have a way to chain these micro-functions so that they run one after the other. Modules may define multiple ante micro-functions and then also provide a default per-module AnteHandler that implements a default, suggested order for these micro-functions. 

Reviewer 3: shouldn't it be 

```go
Combo(bank.SignatureMicroFn, distribution.ValidateMemoMicroFn, ... ).AnteHandler()
```

I'm confused to what `Combo` does/is

Users can order the AnteHandlers easily by simply using the ModuleManager. The ModuleManager will take in a list of AnteHandlers and return a single AnteHandler that runs each AnteHandler in the order of the list provided. If the user is comfortable with the default ordering of each module, this is as simple as providing a list with each module's antehandler (exactly the same as BeginBlocker and EndBlocker).

> If however, users wish to change the order or add, modify, or delete ante micro-functions in anyway; they can always define their own ante micro-functions and add them explicitly to the list that gets passed into module manager.

Reviewer 2: This is what I find to be better over decorators, especially if the chain of decorators in long @ethanfrey.

Reviewer: >  If however, users wish to change the order or add, modify, or delete ante micro-functions in anyway; they can always define their own ante micro-functions and add them explicitly to the list that gets passed into module manager. 

How is this different than decorators?
They are identical in taking a list of functions. The same list.

The difference is that the decorator can run cleanup *after* the call, or modify the result.
It is the same chain of functions in both cases, just the power you give them. 

Thinks like making a sub-transaction after collecting fees (and only committing msg updates on success), or catching panics, are clear places where the chain design doesn't work. How do you plan to handle those?

Reviewer: > If the decorator construction is done in a single place and not convoluted by abstraction, then the readability is the same in both approaches, I agree.

:+1:

I feel the `Module interface` is a very nice standard. But no need to force the `ModuleManager` on everything.

#### Default Workflow:

##### SDK code:

```golang

> func Chainer(order []AnteHandler) AnteHandler {

Reviewer 2: @AdityaSripal how would this handle catching an `OutOfGas` panic?

Author: It doesn't. I think it would still need the ModuleManager AnteHandler to set the GasMeter and have a defer clause before iterating over each individual antehandler function.

>     return func(ctx Context, tx Tx, simulate bool) (newCtx Context, res Result, abort bool) {

Reviewer: Why even have abort?

If you are designing interfaces, maybe try: `(Context, error)` return value - more go-like.
As far as I can see, the result is only used to store a possible error value, as a success result is not passed down the chain, nor is it propagated to the Handlers at the end

Reviewer 2: Yeah, we shouldn't need an `abort` here. Just a `(Context, error)`

Author: Sure will update

    }
}
```

```golang
func VerifySignatures(ctx Context, tx Tx, simulate bool) (newCtx Context, res Result, abort bool) {
    // verify signatures
    // Returns InvalidSignature Result and abort=true if sigs invalid
    // Return OK result and abort=false if sigs are valid
}

func ValidateMemo(ctx Context, tx Tx, simulate bool) (newCtx Context, res Result, abort bool) {
    // validate memo
}

AuthModuleAnteHandler := Chainer([]AnteHandler{VerifySignatures, ValidateMemo})
```

```golang
func DeductFees(ctx Context, tx Tx, simulate bool) (newCtx Context, res Result, abort bool) {
    // Deduct fees from tx
    // Abort if insufficient funds in account to pay for fees
}

func CheckMempoolFees(ctx Context, tx Tx, simulate bool) (newCtx Context, res  Result, abort bool) {
    // If CheckTx: Abort if the fees are less than the mempool's minFee parameter
}

DistrModuleAnteHandler := Chainer([]AnteHandler{CheckMempoolFees, DeductFees})
```

```golang
type ModuleManager struct {
    // other fields
    AnteHandlerOrder []AnteHandler
}

func (mm ModuleManager) GetAnteHandler() AnteHandler {
    retun Chainer(mm.AnteHandlerOrder)
}
```

##### User Code:

```golang

> moduleManager.SetAnteHandlerOrder([]AnteHandler(AuthModuleAnteHandler, DistrModuleAnteHandler))

Reviewer: I know you mention weave as hard to reason - but here, the actual stack is split over multiple modules, making it quite hard to see all the pieces and reason about it.

At the least, the entire flow should be visible in one place, to reduce the error of forgetting or duplicating some check (or placing in the wrong order).

### Decorator Pattern

The [weave project](https://github.com/iov-one/weave) achieves AnteHandler modularity through the use of a decorator pattern. The interface is designed as follows:

```golang
// Decorator wraps a Handler to provide common functionality
// like authentication, or fee-handling, to many Handlers
type Decorator interface {
	Check(ctx Context, store KVStore, tx Tx, next Checker) (*CheckResult, error)
	Deliver(ctx Context, store KVStore, tx Tx, next Deliverer) (*DeliverResult, error)
}
```

Each decorator works like a modularized SDK antehandler function, but it can take in a `next` argument that may be another decorator or a Handler (which does not take in a next argument). These decorators can be chained together, one decorator being passed in as the `next` argument of the previous decorator in the chain. The chain ends in a Router which can take a tx and route to the appropriate msg handler.

A key benefit of this approach is that one Decorator can wrap its internal logic around the next Checker/Deliverer. A weave Decorator may do the following:

> ```golang

Reviewer 3: ```suggestion
```go
```

Reviewer 3: ditto for the rest of the code snippets in this file

    retun Chainer(mm.AnteHandlerOrder)
}
```

##### User Code:

```golang
moduleManager.SetAnteHandlerOrder([]AnteHandler(AuthModuleAnteHandler, DistrModuleAnteHandler))

app.SetAnteHandler(mm.GetAnteHandler())
```

#### Custom Workflow

##### User Code

```golang
func CustomSigVerify(ctx Context, tx Tx, simulate bool) (newCtx Context, err error) {
    // do some custom signature verification logic
}
```

```golang
// Micro-functions allow users to change order of when they get executed, and swap out default ante-functionality with their own custom logic.
// Note that users can still chain the default distribution module handler, and auth micro-function along with their custom ante function
moduleManager.SetAnteHandlerOrder([]AnteHandler(ValidateMemo, CustomSigVerify, DistrModuleAnteHandler))
```

Pros:
1. Allows for ante functionality to be as modular as possible.
2. For users that do not need custom ante-functionality, there is little difference between how antehandlers work and how BeginBlock and EndBlock work in ModuleManager.
3. Still easy to understand

Cons:
1. Cannot wrap antehandlers with decorators like you can with Weave.

### Simple Decorators

> This approach takes inspiration from Weave's decorator design while trying to minimize the number of breaking changes to the SDK and maximizing simplicity.

Reviewer 3: can you explain here in the intro, what's the difference with what's under the `Decorator Pattern` section? 

```
changes to the SDK and maximizing simplicity. The difference(s) is/are ...
```

> # ADR 009: Modular AnteHandler

Reviewer 4: ```suggestion
# ADR 010: Modular AnteHandler
```
sorry for the confusion the title of the PRs had two 008's

Cons:
1. Cannot wrap antehandlers with decorators like you can with Weave.

### Simple Decorators

This approach takes inspiration from Weave's decorator design while trying to minimize the number of breaking changes to the SDK and maximizing simplicity.

Allow Decorator interface that can be chained together to create an SDK AnteHandler.

This allows users to choose between implementing an AnteHandler by themselves and setting it in the baseapp, or use the decorator pattern to chain their custom decorators with SDK provided decorators in the order they wish.

```golang
// A Decorator wraps an AnteHandler, and can do pre- and post-processing on the next AnteHandler
type Decorator interface {
    AnteHandle(ctx Context, tx Tx, simulate bool, next AnteHandler) (newCtx Context, err error)
}
```

```golang
// ChainDecorators will recursively link all of the Decorators in the chain and return a final AnteHandler function
// This is done to preserve the ability to set a single AnteHandler function in the baseapp.

> func ChainDecorators(chain ...Decorator) AnteHandler {

Reviewer: This does decorate the chain of AnteHandlers.
But returns before the Handler is called.

In such a case, the simple ordered-list of ante handlers makes as much sense.
The point of decorators was to wrap the handler execution and be able to do cleanup *after* the *Handler* was called.

If you just want to build a AnteHandler and not call defered code after execution, then the simpler chain approach works well.

Author: This decorates the chain of AnteHandlers so that it is possible for one `AnteDecorator` to wrap and cleanup the `next` antehandler. It is clearly more powerful than the chained micro-function approach since the chained micro-function cannot wrap over each other (So it cannot do defer-cleanup).

It's true, this does not allow users to decorate a `MsgHandler` or wrap the entire handler logic with a decorator. It's a deliberate choice to make decorators capable of modifying the authentication and validation work of `AnteHandler`, since that is where most of the use-case for decorators lie in my opinion.

I don't think there's any reason why a decorator for ModuleA should have the power to modify the results from a `MsgHandler` in ModuleB. It does make sense that a decorator in `ModuleA` may perform additional authentication/validation work on a tx, even if it contains a msg for ModuleB.

    ctx.GasMeter = NewGasMeter(tx.Gas)

    defer func() {
        // recover from OutOfGas panic and handle appropriately
    }

    return next(ctx, tx, simulate)
}

// Signature Verification decorator. Verify Signatures and move on
type SigVerifyDecorator struct{}

func (svd SigVerifyDecorator) AnteHandle(ctx Context, tx Tx, simulate bool, next AnteHandler) (newCtx Context, err error) {
    // verify sigs. Return error if invalid

    // call next antehandler if sigs ok
    return next(ctx, tx, simulate)
}

// User-defined Decorator. Can choose to pre- and post-process on AnteHandler
type UserDefinedDecorator struct{
    // custom fields
}

func (udd UserDefinedDecorator) AnteHandle(ctx Context, tx Tx, simulate bool, next AnteHandler) (newCtx Context, err error) {
    // pre-processing logic

    ctx, err = next(ctx, tx, simulate)

    // post-processing logic
}
```

```golang
// Create final antehandler by chaining the decorators together
antehandler := ChainDecorators(NewSetupDecorator(), NewSigVerifyDecorator(), NewUserDefinedDecorator())

// Set chained Antehandler in the baseapp
bapp.SetAnteHandler(antehandler)
```

Pros:

> 1. Allows one decorator to pre- and post-process the next AnteHandler, similar to the Weave design.

Reviewer: There is no concept of `AnteHandler(); Handler()` in weave.
It is `Decorator(Handler)`.

It is not just the AnteHandler. The enormous body of DeliverTx, runTx and runMsg in BaseApp show all the code that cannot be customized. *Exactly* because there is no other place to put code that runs both before and after all handlers.

Author: I agree all of the functions you mentioned above could do with a refactor that makes them simpler and more readable. I'm not convinced that anything they do should be customizable though. 

Why would a chain decide not to catch OOG panics and return with the appropriate result?
Why would a chain decide not to cache the state and only write state updates if msg-handler passes?

I think all chains will want to do this. If you have an example of some baseapp logic that is hardcoded but should be customizable, please share it in a reply.

I think this comes back to the trade-off between being opinionated and being flexible. Given that every chain will be doing this, I don't see why we should make every user remember to add the relevant decorators to their chain and add that unnecessary burden along with the risk that developers forget to do this.

Given that the code in `runTx` and `DeliverTx` something every SDK user should be running. I think it makes sense to be opinionated here and have it implemented directly into the `DeliverTx` implementation.

This along with my comment [here](https://github.com/cosmos/cosmos-sdk/pull/4942#discussion_r318182836) is why I think it's better to have decorators deliberately restricted to the AnteHandler, rather than have it wrap the entire Handler execution path.
